product_abc_classification/models/xyz_analysis_profile.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2020 ForgeFlow
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from datetime import timedelta
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class XYZAnalysisProfile(models.Model):
+    _name = "xyz.analysis.profile"
+    _description = "XYZ Analysis Profile"
+
+    name = fields.Char()
+    level_ids = fields.One2many(
+        comodel_name="xyz.analysis.profile.level", inverse_name="profile_id"
+    )
+    representation = fields.Char(compute="_compute_representation")
+    data_source = fields.Selection(
+        selection=[("stock_moves", "Stock Moves")],
+        default="stock_moves",
+        string="Data Source",
+        index=True,
+        required=True,
+    )
+    value_criteria = fields.Selection(
+        selection=[("consumption_value", "Consumption Value")],
+        # others: 'sales revenue', 'profitability', ...
+        default="consumption_value",
+        string="Value",
+        index=True,
+        required=True,
+    )
+    past_period = fields.Integer(default=365, string="Days", required=True)
+
+    @api.depends("level_ids")
+    def _compute_representation(self):
+        def _get_sort_key_percentage(rec):
+            return rec.percentage
+
+        for profile in self:
+            profile.level_ids.sorted(key=_get_sort_key_percentage, reverse=True)
+            profile.representation = "/".join(
+                [str(x) for x in profile.level_ids.mapped("display_name")]
+            )
+
+    @api.constrains("level_ids")
+    def _check_levels(self):
+        for profile in self:
+            percentages = profile.level_ids.mapped("percentage")
+            total = sum(percentages)
+            if profile.level_ids and total != 100.0:
+                raise ValidationError(
+                    _("The sum of the percentages of the levels should be 100.")
+                )
+            if profile.level_ids and len({}.fromkeys(percentages)) != len(percentages):
+                raise ValidationError(
+                    _("The percentages of the levels must be unique.")
+                )
+
+    def write(self, vals):
+        return super().write(vals)
+
+    def _fill_initial_product_data(self, date):
+        product_list = []
+        if self.data_source == "stock_moves":
+            return self._fill_data_from_stock_moves(date, product_list)
+        else:
+            return product_list
+
+    def _fill_data_from_stock_moves(self, date, product_list):
+        self.ensure_one()
+        moves = (
+            self.env["stock.move"]
+            .sudo()
+            .search(
```

Comment:
This is going to be very slow. Use search_group, and group by product_id, then accumulating on the field 'product_qty'.